### PR TITLE
InList: fix bug for comparing with Null in the list using the set optimization

### LINKED ISSUE #2800: support decimal in (NULL)

### LINKED ISSUE #2817: IN/NOT IN List: NULL is not equal to NULL
**Describe the bug**

This https://github.com/apache/arrow-datafusion/pull/2156 introduce an optimization for checking value in/not in the list.

But it ignore the `NULL` case.

In the SQL system, NULL is not equal to NULL.

Like below case:
```
select 1 in (2,3,4,5,......100, null)  return null
```

But now It will return false.

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

close #2817 

In this pr: https://github.com/apache/arrow-datafusion/pull/2156, @Ted-Jiang add the set to optimize the Inlist.
But [the implementation](https://github.com/Ted-Jiang/arrow-datafusion/blob/9e0a58aa767010c51e220d2fbb47aa612356b4c6/datafusion/physical-expr/src/expressions/in_list.rs#L208) does not consider the `NULL` case for `IN` or `NOT IN` expr.

In the SQL system like Mysql or spark, NULL is not equal NULL and NULL can't be compared with other data type.
If `A` compare with NULL, the result must be NULL.
```
select 1 not in (2,NULL);
NULL 
select 2 in (1,NULL)
NULL
select NULL in (1,NULL)
NULL
select NULL not in (1,NULL)
NULL
```

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 1]
```diff
@@ -163,16 +139,14 @@ macro_rules! make_contains_primitive {
...
-                    datatype => unimplemented!("Unexpected type {} for InList", datatype),
+                    datatype => unreachable!("InList can't reach other data type {} for {}.", datatype, s),
                 },
                 ColumnarValue::Array(_) => {
                     unimplemented!("InList does not yet support nested columns.")
                 }
             })
             .collect::<Vec<_>>();
-
+        // TODO do we need to replace this below logic by `in_list_primitive`？
```
Comment (liukun4515) [inline datafusion/physical-expr/src/expressions/in_list.rs:149]: Do we need to replace below logic by the macro `collection_contains_check`?

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 2]
```diff
@@ -209,26 +183,109 @@ macro_rules! make_contains_primitive {
...
+        let bool_array = if $NEGATED {
+            // Not in
+            if contains_null {
                 $ARRAY
                     .iter()
-                    .map(|x| x.map(|v| !$LIST_VALUES.contains(&v.try_into().unwrap())))
-                    .collect::<BooleanArray>(),
-            )));
+                    .map(|vop| {
+                        match vop.map(|v| !$SET_VALUES.contains(&v.try_into().unwrap())) {
```
Comment (liukun4515) [inline datafusion/physical-expr/src/expressions/in_list.rs:195]: The f32 is not implemented the `eq` trait, we can't use the `set::<f32>.contains`

Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:195]: We just use the `set::<Scalarvalue::Float32/Float64>

**Comment by viirya** [comment]
I don't get it clearly what this is targeted to fix. The issue it closes is "support decimal in (NULL)", but by a quick look, seems the change is more than that. Although there is a few description, but I cannot get it from it too. 

Could you rephrase the issue this is going to fix?

**Comment by liukun4515** [comment]
> I don't get it clearly what this is targeted to fix. The issue it closes is "support decimal in (NULL)", but by a quick look, seems the change is more than that. Although there is a few description, but I cannot get it from it too.
> 
> Could you rephrase the issue this is going to fix?

sorry for the confused description.

I have described the issue in https://github.com/apache/arrow-datafusion/issues/2817.

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 3]
```diff
@@ -1139,4 +1242,192 @@ mod tests {
...
+            lit(ScalarValue::Boolean(None)),
+        ];
+        for _ in 1..(OPTIMIZER_INSET_THRESHOLD + 1) {
+            list.push(lit(ScalarValue::Boolean(Some(true))));
+        }
+        in_list!(
+            batch,
+            list.clone(),
+            &false,
+            vec![Some(true), None, None],
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/in_list.rs:1265]: 👍

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 4]
```diff
@@ -1139,4 +1242,192 @@ mod tests {
...
+    }
+
+    #[test]
+    fn in_list_set_float64() -> Result<()> {
+        let schema = Schema::new(vec![Field::new("a", DataType::Float64, true)]);
+        let a = Float64Array::from(vec![Some(0.0), Some(2.0), None]);
+        let col_a = col("a", &schema)?;
+        let batch = RecordBatch::try_new(Arc::new(schema.clone()), vec![Arc::new(a)])?;
+
+        // expression: "a in (0.0,3.0,4.0 ....)"
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/in_list.rs:1325]:
```suggestion
        // expression: "a in (0.0,Null,3.0,4.0 ....)"
```

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 5]
```diff
@@ -1139,4 +1242,192 @@ mod tests {
 
         Ok(())
     }
+
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/in_list.rs:1245]: I really like the test coverage

File: datafusion/physical-expr/src/expressions/in_list.rs (hunk 2 above)
Comment (alamb) [reply datafusion/physical-expr/src/expressions/in_list.rs:195]: Elsewhere in DataFusion (including in `ScalarValue`) we use `ordered_float` to compare floating point numbers

It might be possible  to use `set::<OrderedFloat<f32>>`, which would be more space efficient (fewer bytes than `ScalarValue`) as well as faster (as the comparison doens't have to dispatch on the type each time)


https://github.com/apache/arrow-datafusion/blob/88b88d4360054a85982987aa07b3f3afd2db7d70/datafusion/common/src/scalar.rs#L33

File: datafusion/physical-expr/src/expressions/in_list.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-expr/src/expressions/in_list.rs:149]: I reviewed the logic below and it looked correct to me -- are you asking about the trying to refactor to reduce duplication?

**Review by alamb — APPROVED** [review]
Thank you @liukun4515  -- I reviewed the tests carefully and I think this looks good.

The code "feels" a little more complicated than needed but I think it is doing what it needs to and the test coverage is good.

I wonder if it is worth just 1 test in `sql_integration` somewhere that ensures this implementation is hooked up correctly rather than just relying on tests in `in_list.rs`

I think it would be good if @Ted-Jiang  was also able to review this PR prior to merging it, but it isn't necessary.

File: datafusion/physical-expr/src/expressions/in_list.rs (hunk 1 above)
Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:149]: > I reviewed the logic below and it looked correct to me -- are you asking about the trying to refactor to reduce duplication?

Yes, I want to replace the logic in the `in_list_primitive` by the macro `collection_contains_check`.
I will replace the duplicated code in the in_list_primitive.

File: datafusion/physical-expr/src/expressions/in_list.rs (hunk 2 above)
Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:195]: @alamb  I will optimize this with follow-up pr with issue https://github.com/apache/arrow-datafusion/issues/2831

**Comment by liukun4515** [comment]
> Thank you @liukun4515 -- I reviewed the tests carefully and I think this looks good.
> 
> The code "feels" a little more complicated than needed but I think it is doing what it needs to and the test coverage is good.
> 
> I wonder if it is worth just 1 test in `sql_integration` somewhere that ensures this implementation is hooked up correctly rather than just relying on tests in `in_list.rs`
> 
> I think it would be good if @Ted-Jiang was also able to review this PR prior to merging it, but it isn't necessary.

https://github.com/apache/arrow-datafusion/issues/2832 issue to track it

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 6]
```diff
@@ -1139,4 +1242,192 @@ mod tests {
...
+        let a = BooleanArray::from(vec![Some(true), None, Some(false)]);
+        let col_a = col("a", &schema)?;
+        let batch = RecordBatch::try_new(Arc::new(schema.clone()), vec![Arc::new(a)])?;
+
+        // expression: "a in (true,null,true.....)"
+        let mut list = vec![
+            lit(ScalarValue::Boolean(Some(true))),
+            lit(ScalarValue::Boolean(None)),
+        ];
+        for _ in 1..(OPTIMIZER_INSET_THRESHOLD + 1) {
```
[the remark this answers is not in the document]
Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:1258]: change to `0..OPTIMIZER_INSET_THRESHOLD`

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 7]
```diff
@@ -363,25 +349,28 @@ fn make_set_contains_decimal(
...
-            array
-                .iter()
-                .map(|v| v.map(|v128| !native_set.contains(&v128)))
-                .collect::<BooleanArray>()
-        }
-    }
+    collection_contains_check!(array, native_set, negated, contains_null)
+}
+
+fn set_contains_utf<OffsetSize: OffsetSizeTrait>(
```
Comment (viirya) [inline datafusion/physical-expr/src/expressions/in_list.rs:355]:
```suggestion
fn make_set_contains_utf8<OffsetSize: OffsetSizeTrait>(
```

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 8]
```diff
@@ -532,66 +521,131 @@ impl PhysicalExpr for InListExpr {
             match value_data_type {
                 DataType::Boolean => {
                     let array = array.as_any().downcast_ref::<BooleanArray>().unwrap();
-                    set_contains_with_negated!(array, set, self.negated)
+                    Ok(set_contains_for_primitive!(
```
Comment (viirya) [inline datafusion/physical-expr/src/expressions/in_list.rs:524]: I'm wondering why `set_contains_for_primitive` is macro, but `set_contains_utf` and `make_set_contains_decimal` are methods?

**Review by viirya — APPROVED** [review]
Looks good to me. The logic handling null looks correct to me. I have a few comments about method naming and macro/method.

File: datafusion/physical-expr/src/expressions/in_list.rs (hunk 8 above)
Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:524]: `set_contains_utf` and `make_set_contains_decimal` are just apply to a specified data type case which are not compatible with other primitive case.
`set_contains_for_primitive` will apply to many similar case and same logic

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 9]
```diff
@@ -170,86 +146,114 @@ macro_rules! make_contains_primitive {
...
     }};
 }
 
-macro_rules! set_contains_with_negated {
-    ($ARRAY:expr, $LIST_VALUES:expr, $NEGATED:expr) => {{
-        if $NEGATED {
-            return Ok(ColumnarValue::Array(Arc::new(
+macro_rules! collection_contains_check {
+    ($ARRAY:expr, $VALUES:expr, $NEGATED:expr, $CONTAINS_NULL:expr) => {{
+        let bool_array = if $NEGATED {
```
Comment (viirya) [inline datafusion/physical-expr/src/expressions/in_list.rs:220]: Two `$NEGATED` cases look very similar, is it possible to combine them to save code?

e.g.

```rust
if $CONTAINS_NULL {
  $ARRAY
    .iter()
    .map(|vop| match vop.map(|v| {
      if $NEGATED { !$VALUES.contains(&v) } else { $VALUES.contains(&v) }
    }) {
      Some(true) if $NEGATED => None,
      Some(false) if !$NEGATED => None,
      x => x,
    })
  .collect::<BooleanArray>()
} else {
  $ARRAY
    .iter()
    .map(|vop| vop.map(|v| {
      if $NEGATED { !$VALUES.contains(&v) } else { $VALUES.contains(&v) }
    })
    .collect::<BooleanArray>()
}
```

Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:220]: Thanks @viirya 
these two branches can be merged.

Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:220]: @viirya @alamb 
each loop will check the `$NEGATED` and has two branches for switching.
Is this can use the vectorization?
I am not sure about this, so I will remain current implementation.
We can discuss it until getting the conclusion.

Comment (liukun4515) [reply datafusion/physical-expr/src/expressions/in_list.rs:220]: https://github.com/apache/arrow-datafusion/issues/2833 track this.

Comment (alamb) [reply datafusion/physical-expr/src/expressions/in_list.rs:220]: We have definitely seen cases where

```rust
if $NEGATED {
  // loop
} else {
  // loop
}
```

Was faster than 

```rust
loop {
  if $NEGATED {
    // ..
  } else {
   // ..
  } 
}
```

But I think the only way to find out if that applies in this case would be with a benchmark
